src/rocksdb_store.rs

```diff
@@ -0,0 +1,110 @@
+    #[error("MessagePack deserialization error")]
+    MessagePack(#[from] rmp_serde::decode::Error),
+    /// The value for the given key was not found
+    #[error("No value found for the given key")]
+    NotFound,
+}
+
+/// Errors that can occur during `PkvStore::set`
+#[derive(thiserror::Error, Debug)]
+pub enum SetError {
+    /// An internal error from the rocksdb crate
+    #[error("Rocksdb error")]
+    Rocksdb(#[from] rocksdb::Error),
+    /// Error when serializing the value
+    #[error("MessagePack serialization error")]
+    MessagePack(#[from] rmp_serde::encode::Error),
+    /// Invalid utf8 key name
+    #[error("Invalid RocksDB key name error")]
+    RocksdbKeyName(#[from] Utf8Error),
+}
+
+impl RocksDBStore {
+    pub(crate) fn new(config: &StoreConfig) -> Self {
+        let mut options = rocksdb::Options::default();
+        options.set_error_if_exists(false);
+        options.create_if_missing(true);
+        options.create_missing_column_families(true);
+
+        let dirs = ProjectDirs::from(
+            config.qualifier.as_deref().unwrap_or(""),
+            &config.organization,
+            &config.application,
+        );
+
+        let parent_dir = match dirs.as_ref() {
+            Some(dirs) => dirs.data_dir(),
+            None => Path::new("."), // todo: maybe warn?
+        };
+
+        let db_path = parent_dir.join("bevy_rocksdb_pkv");
+
+        let db = rocksdb::DB::open(&options, db_path).expect("Failed to init key value store");
+
+        Self { db }
+    }
+}
+
+impl StoreImpl for RocksDBStore {
+    type GetError = GetError;
+    type SetError = SetError;
+
+    /// Serialize and store the value
+    fn set<T: Serialize>(&mut self, key: &str, value: &T) -> Result<(), Self::SetError> {
+        let mut serializer = rmp_serde::Serializer::new(Vec::new()).with_struct_map();
+        value.serialize(&mut serializer)?;
+        self.db.put(&key, serializer.into_inner())?;
+
+        Ok(())
+    }
+
+    /// More or less the same as set::<String>, but can take a &str
+    fn set_string(&mut self, key: &str, value: &str) -> Result<(), Self::SetError> {
+        let bytes = rmp_serde::to_vec(value)?;
+        self.db.put(key, bytes)?;
+
+        Ok(())
+    }
+
+    /// Get the value for the given key
+    /// returns Err(GetError::NotFound) if the key does not exist in the key value store.
+    fn get<T: DeserializeOwned>(&self, key: &str) -> Result<T, Self::GetError> {
+        let bytes = self.db.get(key)?.ok_or(Self::GetError::NotFound)?;
+        let value = rmp_serde::from_slice(&bytes)?;
+        Ok(value)
+    }
+
+    /// Clear all keys and their values
+    /// The RocksDB adapter uses an iterator to achieve this, unlike sled
+    fn clear(&mut self) -> Result<(), Self::SetError> {
+        let kv_iter = self.db.iterator(rocksdb::IteratorMode::Start);
+
+        for kv in kv_iter {
+            let result = kv?;
+            let key = result.0;
+            let key = std::str::from_utf8(&key)?;
```

Comment:
Is this necessary? Isn't the key just going to get converted back into a u8 slice?

Reply:
Line 104 is not necessary, your right. I don't know where my brain was with that one.

Reply:
Also, if we remove `let key = std::str::from_utf8(&key)?;` we can also remove:

```rust
#[error("Invalid RocksDB key name error")]
RocksdbKeyName(#[from] Utf8Error),
```

Reply:
I went ahead and removed it